solutions/src/main/java/ru/sberbank/school/task02/ExtendedFxConversionServiceImpl.java

```diff
@@ -0,0 +1,173 @@
+            if (beneficiary == Beneficiary.BANK) {
+                targetQuote = getLastConcurrency(quoteList, amount, ClientOperation.BUY);
+            } else {
+                targetQuote = getFirstConcurrency(quoteList, amount, ClientOperation.BUY);
+            }
+        } else {
+            if (beneficiary == Beneficiary.BANK) {
+                targetQuote = getFirstConcurrency(quoteList, amount, ClientOperation.SELL);
+            } else {
+                targetQuote = getLastConcurrency(quoteList, amount, ClientOperation.SELL);
+            }
+        }
+
+        if (targetQuote == null) {
+            return Optional.empty();
+        }
+
+        BigDecimal price = BigDecimal.ONE.divide(
+                operation == ClientOperation.BUY ? targetQuote.getOffer() : targetQuote.getBid(),
+                10, BigDecimal.ROUND_HALF_UP);
+
+
+        return Optional.of(price);
+    }
+
+    /**
+     * Возвращает значение цены единицы котируемой валюты для указанного объема.
+     *
+     * @param operation   вид операции
+     * @param symbol      Инструмент
+     * @param amount      Объем
+     * @param delta       допустимое отклонение при отсутствии точного попадания в котируемой валюте
+     * @param beneficiary В чью пользу осуществляется округление
+     * @return Цена для указанного объема
+     */
+    @Override
+    public Optional<BigDecimal> convertReversed(ClientOperation operation,
+                                                Symbol symbol,
+                                                BigDecimal amount,
+                                                double delta,
+                                                Beneficiary beneficiary) {
+        return Optional.empty();
+    }
+
+    private Quote getLastConcurrency(List<Quote> quoteList, BigDecimal amount, ClientOperation operation) {
+
+        List<Quote> listCopy = new ArrayList<>();
+        Collections.copy(listCopy,quoteList);
+        listCopy.sort((o1, o2) -> {
+            if (o1.isInfinity()) {
+                return -1;
+            } else if (o2.isInfinity()) {
+                return 1;
+            } else {
+                return o2.getVolumeSize().compareTo(o1.getVolumeSize());
+            }
+        });
+        return getFirstConcurrency(listCopy, amount, operation);
+    }
+
+    private Quote getFirstConcurrency(List<Quote> quoteList, BigDecimal amount, ClientOperation operation) {
+        int listSize = quoteList.size();
+        if (operation == ClientOperation.BUY) {
+            if (quoteList.get(0).getOffer().multiply(quoteList.get(0).getVolumeSize()).compareTo(amount) > 0) {
+                return quoteList.get(0);
+            }
+            for (int i = 1; i < listSize - 1; i++) {
+                if ((quoteList.get(i).getOffer().multiply(quoteList.get(i).getVolumeSize()).compareTo(amount) > 0)
+                        && (quoteList.get(i).getOffer().multiply(quoteList.get(i - 1).getVolumeSize()).compareTo(amount) < 0)) {
```

Comment:
Line is longer than 120 characters (found 128).

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-648750258>*